cli/flox-rust-sdk/src/models/lockfile.rs

```diff
@@ -902,17 +1038,133 @@ impl LockedManifestCatalog {
+        let mut flake_installables = Vec::new();
+
+        for (install_id, manifest_descriptor) in manifest.install.iter() {
+            let manifest_descriptor = match manifest_descriptor {
+                ManifestPackageDescriptor::FlakeRef(f) => f,
+                _ => continue,
+            };
+
+            let systems = manifest
+                .options
+                .systems
+                .as_deref()
+                .unwrap_or(&*DEFAULT_SYSTEMS_STR);
+
+            for system in systems {
+                flake_installables.push(FlakeInstallableToLock {
+                    install_id: install_id.clone(),
+                    descriptor: manifest_descriptor.clone(),
+                    system: system.to_owned(),
+                });
+            }
+        }
+
+        flake_installables.into_iter()
+    }
+
+    fn split_locked_flake_installables(
+        installables: impl IntoIterator<Item = FlakeInstallableToLock>,
+        seed_lockfile: Option<&LockedManifestCatalog>,
+    ) -> (Vec<LockedPackage>, Vec<FlakeInstallableToLock>) {
+        // todo: consider computing once and passing a reference to the consumer functions.
+        //       we now compute this 3 times during a single lock operation
+        let seed_locked_packages = seed_lockfile.map_or_else(HashMap::new, Self::make_seed_mapping);
+
+        let by_id = installables.into_iter().group_by(|i| i.install_id.clone());
+
+        let (already_locked, to_lock): (Vec<Vec<LockedPackage>>, Vec<Vec<FlakeInstallableToLock>>) =
+            by_id.into_iter().partition_map(|(_, group)| {
+                let unlocked = group.collect::<Vec<_>>();
+                let mut locked = Vec::new();
+
+                for installable in unlocked.iter() {
+                    let Some((locked_descriptor, in_lockfile @ LockedPackage::Flake(_))) =
+                        seed_locked_packages
+                            .get(&(installable.install_id.as_str(), &installable.system))
+                    else {
+                        return Either::Right(unlocked);
+                    };
+
+                    if ManifestPackageDescriptor::from(installable.descriptor.clone())
+                        .invalidates_existing_resolution(locked_descriptor)
+                    {
+                        return Either::Right(unlocked);
+                    }
+
+                    locked.push((*in_lockfile).to_owned());
+                }
+                Either::Left(locked)
+            });
+
+        let already_locked = already_locked.into_iter().flatten().collect();
+        let to_lock = to_lock.into_iter().flatten().collect();
+
+        (already_locked, to_lock)
+    }
+
+    /// Lock a set of flake installables and return the locked packages.
+    /// Errors are collected into [ResolutionFailures] and returned as a single error.
+    ///
+    /// This is the eequivalent to
+    /// [catalog::ClientTrait::resolve] >>= [Self::locked_packages_from_resolution]
```

Comment:
I'm not sure what this is saying. What is this operator?

Reply:
map/process the output of resolve with locked_packages_from_resolution. could be a pipe as well in shell world

yeah lets get rid of the operator :D